---
title: "16S"
date:  "`r format(Sys.time(), 'Last updated: %d/%m/%y')`"
output: html_document
---
#phyloseq
## load
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F)


#install.packages("metagenomeSeq")
#install.packages("phyloseq")
#install.packages("DESeq2")
#install.packages("S4Vectors")
#install.packages('EnhancedVolcano')
#install.packages('microbiome')


library(readr)
library(plyr)
library(dplyr)
library(tidyr)
library(taxa)
#library(metagenomeSeq)
library(openxlsx)
library(readxl)
#library(car)
library(tibble)
library(stringr)
library(gplots)
library(ggplot2)
library(ggthemes)
library(gridGraphics)
library(pheatmap)
#library(DataExplorer)
#library(caret)
#library(lmerTest)
library(phyloseq)
library(vegan)
library(RColorBrewer)
library(DESeq2)
library(factoextra)
library(PairedData)
library(EnhancedVolcano)
library(reshape2)
library(RColorBrewer) # Load a package giving more colors
library(pheatmap) # load a package for making heatmaps
library(grid)
library(gridExtra)
library(ggpubr)
library(microbiome)
library(corrplot)
library(psych)
#library(biomaRt)

#source("RNAseq_functions.R")
source("phyloseq_functions.R")
```

##environment
remove the first line & replace "#" with \t before!!!!!!!!!!!!!!! FOR otu & taxa

```{r}
set<-batch <- "394751-ON" #303431 260926 
#treat <- "IBS" 
#classif<-"_GG_nb_97" # _GG_nb_97 _vs_99  _SL138_nb_97
#RNA_corr<- FALSE


metadata_dir<- paste0("C:/Users/ariel.DESKTOP-6TR9210/Okun Lab Dropbox/Yissachar Lab/Personal folders/Ariel Simon/Thesis/16S_chemo394751")
#if (grepl("efrat",getwd())) 
 # metadata_dir<- paste0("C:/Users/efrat/Google Drive/Microbaiom LAB/",
#                                                      treat,"/16S_",treat,"_",set)
                         
#if (grepl("amirv",getwd())) 
 # metadata_dir<- paste0("C:/Users/amirv/Google Drive/Microbaiom LAB/",
 #                                                     treat,"/16S_",treat,"_",set)
  
result_dir <- paste0(metadata_dir,"/_results_", 
                     Sys.Date(), "_",format(Sys.time(), "%s"))
if (!file.exists(result_dir)) {dir.create(result_dir)}


```


#phyloseq
## shortcut
```{r}
#if phyloseq object already exists, load it
#if (file.exists(paste0(metadata_dir,"/physeq_rarefied_and_scaled",classif,"_",set,".Rdata"))) 
 # {load(paste0(metadata_dir,"/physeq_rarefied_and_scaled",classif,"_",set,".Rdata"))}
```

## preperations
https://joey711.github.io/phyloseq/index.html 
https://gist.github.com/erictleung/eda33bceceebb190eb9a44c93a077d32 
build phyloseq object that includes OTU+TAX+META+tree
```{r fig.height=25, fig.width=12}
#remove the first line & replace "#" with \t before!!!!!!!!!!!!!!! FOR otu & taxa
otu<- read.table(paste0(metadata_dir,"/otu_table.txt"), header = T,  sep= "\t")
taxa<- read.table(paste0(metadata_dir,"/taxonomy.tsv"), header = T,  sep= "\t")
meta_16S_org <- read.table(paste0(metadata_dir,"/meta-data.tsv"), header = T,  sep = "\t")
phy_tree <- read_tree(paste0(metadata_dir,"/tree.nwk"))
#clinical <- read_excel(paste0(metadata_dir,"/clinical_AraC.xlsx"))


# split taxon column and create short_title
taxa_16S_df<-prep_taxa(taxa)



meta_16S <- meta_16S_org %>%
              mutate(batch= str_split_fixed(sample,"\\-",4)[,1]) %>%
              mutate(SampleID=str_replace(sample,paste0(set,"-"),""))  %>% # remove prefix
              mutate(SampleID=str_replace_all(sample,"-","_"))  %>% # replace to underscore
              mutate(sample_name=paste0(sample,"_",classification,"_",patient)) # add TimeW+PatientID


#clinical$TimeW <- meta_16S[match(clinical$SampleID, meta_16S$SampleID), "TimeW"]

otu_df <- otu %>% 
              as.data.frame(stringsAsFactors=FALSE)  %>%
              column_to_rownames("OTU.ID") %>% 
              t()  %>% 
              as.data.frame(stringsAsFactors=FALSE)  %>%
              rownames_to_column("SampleID")  %>%
              mutate(SampleID=str_replace(SampleID,paste0("X",set,"."),""))  %>% # remove prefix
              mutate(SampleID=str_replace(SampleID,"X",""))  %>% # replace dot with underscore
              mutate(SampleID=str_replace(SampleID,"\\.","-"))  %>% # replace dot with underscore
              mutate(SampleID=meta_16S[match(SampleID,meta_16S$sample), "sample_name"]) %>% #+TimeW+PatientID
              column_to_rownames("SampleID") %>% 
              t()  %>% 
              as.data.frame(stringsAsFactors=FALSE)  

head(otu_df)
rownames(meta_16S) <- meta_16S$sample_name
#meta_16S$SampleID <- NULL

#clinical$SampleID <- paste0(clinical$SampleID,"_", substr(clinical$TimeW,1,1))

OTU = otu_table(as.matrix(otu_df), taxa_are_rows = TRUE)
TAX = tax_table(as.matrix(taxa_16S_df))
META <- sample_data(meta_16S)


#sanity taxa names
length(taxa_names(OTU))
length(taxa_names(TAX))
length(taxa_names(phy_tree))

head(taxa_names(OTU))
head(taxa_names(TAX))

#sanity sample names
sample_names(OTU)
sample_names(META)

openxlsx::write.xlsx (taxa_16S_df, paste0(metadata_dir,"/taxa_16S_df.xlsx"),
                              sheetName ="taxa_16S_df" ,colNames = TRUE,rowNames=TRUE) 

```

## build phyloseq object

```{r}
physeq = phyloseq(OTU, TAX, META, phy_tree)
physeq
sample_data(physeq)
rank_names(physeq)
sample_variables(physeq)
nsamples(physeq)
head(otu_table(physeq))
```
## clean the empty, white-space, NA values
not relevant as all the rest is done baset on short_title
```{r}
physeq
taxa_mtrx <- as(access(physeq, "tax_table"), "matrix")
names(taxa_mtrx)<-taxa_names(physeq)

bad_empty=c(NA, "", " ", "\t", "signed", "k__Bacteria")
print(length(names(taxa_mtrx)))
keep_species <- names(taxa_mtrx)[ !(taxa_mtrx %in% bad_empty) ]
# names(taxa_mtrx)[ (taxa_mtrx %in% bad_empty) ]
print(length(keep_species))

physeq_clean <- prune_taxa(keep_species, physeq)
physeq_clean

#taxa_mtrx_clean <- as(access(physeq_clean, "tax_table"), "matrix")
#nrow(taxa_mtrx_clean)

#Keep only OTUs that appear in more than 1 samples
#physeq_clean = filter_taxa(physeq_clean, function(x) sum(x>0)>1,TRUE)
#physeq_clean

```

## clean the samples 

```{r}
physeq_clean
minTotRelAbun <- 3e-5
physeq_taxa_sums <- taxa_sums(physeq_clean)
physeq_taxa_sums

#data is first transformed to relative abundance,then filtered that only OTUs with a mean greater than 10^-5 are kept
part  <- physeq_taxa_sums/sum(physeq_taxa_sums)
qplot(part, log = "x") +
    geom_vline(xintercept = minTotRelAbun) # possible cutoff

keepTaxa <- taxa_names(physeq_clean)[which((physeq_taxa_sums/sum(physeq_taxa_sums)) > minTotRelAbun)]
physeq_pruned <- prune_taxa(keepTaxa, physeq_clean)
physeq_pruned


#abundances greater than 30% and appear in 20% of the samples
#if (batch=="303431"){
 # f1<- filterfun_sample(topp(0.4)) 
#  to_filter <- genefilter_sample(physeq_pruned, f1, A=(1/10*nsamples(physeq_clean)))
#}else{
#  f1<- filterfun_sample(topp(0.3)) 
#  to_filter <- genefilter_sample(physeq_pruned, f1, A=(1/30*nsamples(physeq_clean)))
# }
# sum(to_filter)
# physeq_filtered <- prune_taxa(to_filter, physeq_pruned)
# physeq_filtered
```



##take same number of OTUs with rarefy_even_depth
https://micca.readthedocs.io/en/latest/phyloseq.html 
```{r fig.height=8, fig.width=10}

set.seed(123)
# Plot the rarefaction curves using vegan function rarecurve()
rarecurve(t(otu_table(physeq_filtered)), step=50, cex=0.5)
#rarecurve(t(otu_table(physeq_clean)), step=50, cex=0.5)

#Rarefy the samples without replacement. Rarefaction is used to simulate even number of reads per sample. In this example, the rarefaction depth chosen is the 80% of the minimum sample depth in the dataset (in this case ? reads per sample).

sample_sums(physeq_filtered)
# take the same number of speciman from each sample
physeq_rarefied = rarefy_even_depth(physeq_filtered, rngseed=123,
                                    sample.size=0.99*min(sample_sums(physeq_pruned)), replace=T) #physeq_pruned is # physeq_filtered
physeq_rarefied
sample_sums(physeq_rarefied)
rarecurve(t(otu_table(physeq_rarefied)), step=50, cex=0.5)
dev.copy(jpeg,filename=paste0(result_dir,"/rarefied_",set, ".jpg"), 
                 width=1000, height=700)
dev.off()
      
sample_data(physeq_rarefied)
rank_names(physeq_rarefied)
sample_variables(physeq_rarefied)

physeq_rarefied_data <- psmelt(physeq_rarefied)
str(physeq_rarefied_data)

openxlsx::write.xlsx (physeq_rarefied_data,
                              paste0(metadata_dir,"/physeq_rarefied_data.xlsx"),
                              sheetName ="physeq_rarefied_data" ,col.names = TRUE) 

sample_data(physeq_rarefied)$classification <- as.factor(sample_data(physeq_rarefied)$classification)

openxlsx::write.xlsx (otu_table(physeq_rarefied),paste0(metadata_dir,"/physeq_rarefied_otu_table.xlsx"),
                              sheetName ="physeq_rarefied_otu_table.xlsx" ,col.names = TRUE,row.names=TRUE) 

qplot(log10(rowSums(otu_table(physeq_rarefied)))) +
  xlab("Logged counts-per-sample")
```



##Scale taxa to relative abundance (of defined rank)
converts the counts from each sample into their frequencies, meaning as proportions or relative abundances. 
```{r}
#transform to relative abundance
physeq_rarefied_scaled = 
  transform_sample_counts(physeq_rarefied, function(y) 100 * y/sum(y))
physeq_rarefied_scaled
#sample_data(physeq_rarefied_scaled)$TimeW <- as.factor(sample_data(physeq_rarefied_scaled)$TimeW)

#``melted'' phyloseq data is stored much less efficiently, usually for producing graphics
physeq_rarefied_scaled_data <- psmelt(physeq_rarefied_scaled)
str(physeq_rarefied_scaled_data)

openxlsx::write.xlsx (physeq_rarefied_scaled_data,
                              paste0(metadata_dir,"/physeq_rarefied_scaled_data.xlsx"),
                              sheetName ="physeq_rarefied_scaled_data" ,
                              row.names = TRUE, col.names = TRUE, append = FALSE) 


save(meta_16S,
     taxa_16S_df,
     physeq_rarefied,
     physeq_rarefied_data,
     physeq_rarefied_scaled,physeq_rarefied_scaled_data,
     file=paste0(metadata_dir,"/physeq_rarefied_and_scaled_",set,".Rdata"))

qplot(log10(rowSums(otu_table(physeq_rarefied_scaled)))) +
  xlab("Logged scaled counts-per-sample")
```



## merge the OTUs at the each rank=level and plot. 
Given a taxonomic rank , the phyloseq function tax_glom merges the OTUs with the same taxonomy, summing the abundances.
Usually, there will be multiple OTUs with the same taxonomy assignment. These may be spurious (perpetuated errors, chimeras) or could be different species where the sequence identity is less than 97% but the genus-level taxonomy is the same. Some OTUs may also contain multiple species if they are more than 97% identical. 
```{r fig.height=8, fig.width=10}
mpt <- merge_plot_taxa(physeq_rarefied,
                       physeq_rarefied_scaled,
                       physeq_rarefied_data,
                       physeq_rarefied_scaled_data,
                       field= "classification",
                       field2= "patient")


```


### plot network
```{r fig.height=12, fig.width=12}
theme_set(theme_bw())

ig <- make_network(physeq_rarefied_scaled, "taxa", max.dist=0.5)
plot_network(ig, physeq_rarefied_scaled, "taxa", color = "short_title", line_weight = 0.4, label = NULL)

dev.copy(jpeg,filename=paste0(result_dir,"/taxa_network_plot", ".jpg"), 
           width=1500, height=1500)
dev.off()

```


## Alpha diversity
calculated on rarified but not scaled data
```{r}

wb <- openxlsx::createWorkbook()
posStyle <- createStyle(fontColour = "#006100", fgFill = "#C6EFCE")
s <- createStyle(numFmt = "$ #,##0.00")
sheet_name <- "alpha_diversity"
addWorksheet(wb, sheet_name)

alpha_diversity = estimate_richness(physeq_rarefied)
#alpha_diversity$ShannonEvenness = rich$Shannon / log(rich$Observed)
#rownames(alpha_diversity) = gsub("X", "", rownames(alpha_diversity))
writeData(wb,sheet_name , alpha_diversity, colNames = TRUE) 
setColWidths(wb, sheet_name, cols = 1:8, widths = "auto")

measures <- c("Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", "Fisher")
plot_richness(physeq_rarefied, x="classification", measures=measures) + 
  geom_boxplot()+
  ggtitle("Richness box per classification")
dev.copy(jpeg,filename=paste0(result_dir,"/plot_box_Alpha_classification", ".jpg"), 
           width=1500, height=1000)
dev.off()



#Test whether the observed number of OTUs differs significantly between groups

#m<-"Observed"
sheet_name <- "pairwise_wilcox_test_all"
addWorksheet(wb, sheet_name)

wilcox <- data.frame(measure=character(),TimeW_pv=double(),treat1=character(),treat2=character(),treat3=character())
for (m in measures){
  wilcox_classification <- pairwise.wilcox.test(alpha_diversity[[m]], sample_data(physeq_rarefied)$classification )
  wilcox <- rbind(wilcox, cbind(Mmeasure = m, 
                                class_pv= as.double(wilcox_classification$p.value),
                                treat1= colnames(as.data.frame(wilcox_classification$p.value)),
                                treat2= colnames(as.data.frame(wilcox_classification$p.value)),
                                treat3= rownames(as.data.frame(wilcox_classification$p.value))))
  print(as.data.frame(wilcox_classification$p.value))
}
wilcox


addStyle(wb, 1, style = s, cols=2:5, rows=2:nrow(wilcox), gridExpand = TRUE)
## set a default number format for numeric columns of data.frames
options("openxlsx.numFmt" = "0.00")
writeData(wb,sheet_name , wilcox, colNames = TRUE) 
conditionalFormatting(wb, sheet_name, cols=2:5, rows=2:nrow(wilcox),rule="B2<0.05", style = posStyle)
setColWidths(wb, sheet_name, cols = 1:8, widths = "auto")


```

## Beta diversity
Plot PCoA bray as distance & calculate ANOVA
ANOVA-Test whether the phases differ significantly from each other using the permutational ANOVA (PERMANOVA) analysis
ANOVA significance tests are performed using F-tests based on sequential sums of squares from permutations of the raw data.


```{r}

sheet_name <- "Beta_diversity_anova"
addWorksheet(wb, sheet_name)

# PCoA plot using the unweighted UniFrac as distance
theme_set(theme_bw())
# methods for distance

#dist_methods <- c("bray")
dst<- "bray"
ord_meths = c("DCA", "CCA", "RDA", "DPCoA", "NMDS", "MDS", "PCoA")
dist <- phyloseq::distance(physeq_rarefied, method=dst, weighted=T) #, weighted=T

# ANOVA-Test whether the group: TimeW,PatientID  differ significantly from each
anova_all <- adonis2(dist ~ sample_data(physeq_rarefied_scaled)$classification*sample_data(physeq_rarefied_scaled)$patient)
anova_all
writeData(wb,sheet_name , anova_all, colNames = TRUE, rowNames = TRUE) 

anova_all2 <- adonis2(dist ~ sample_data(physeq_rarefied_scaled)$patient*sample_data(physeq_rarefied_scaled)$classification)
anova_all2
writeData(wb,sheet_name , anova_all2, colNames = TRUE, rowNames = TRUE, startRow = 10) 

# ANOVA-Test whether the group: TimeW,  differ significantly from each
anova_time <- adonis2(dist ~ sample_data(physeq_rarefied_scaled)$classification )
anova_time
writeData(wb,sheet_name , anova_time, colNames = TRUE, rowNames = TRUE, startRow = 20) 

# ANOVA-Test whether the group: patientID,  differ significantly from each
anova_pid <- adonis2(dist ~ sample_data(physeq_rarefied_scaled)$patient )
anova_pid
writeData(wb,sheet_name , anova_pid, colNames = TRUE, rowNames = TRUE, startRow = 25) 

setColWidths(wb, sheet_name, cols = 1:8, widths = "auto")
openxlsx::saveWorkbook(wb, file=paste0(result_dir,"/diversity_statistics_", batch, ".xlsx"), 
                       overwrite = TRUE)


for (mth in ord_meths){
#  for (dst in dist_methods){
    #mth<- "DCA"
      
    print(paste("method:", mth, "dist method: ", dst))

    ord <- ordinate(physeq_rarefied, method=mth, distance=dst)

    #plots
    print(plot_ordination(physeq_rarefied, ord, color="classification", label = "SampleID") +
      theme(aspect.ratio=1) +
      geom_point(size = 5) +
  #need >3 points to calculate an ellipse
    #stat_ellipse(geom = "circle") +coord_fixed()+
    #geom_path() +coord_fixed()+
      ggtitle(paste( dst, "Distance Method,",mth,"Ordination Method",
                     "\n ANOVA:",anova$aov.tab$`Pr(>F)`[1]))+ 
        scale_shape_manual(values = c(20,18))+
        theme(plot.title = element_text(size = 16, face = "bold")))
    dev.copy(jpeg,filename=paste0(result_dir,"/plot_Beta_",mth,"_class", ".jpg"), 
               width=600, height=800)
    dev.off()
    print(plot_scree(ordinate(physeq_rarefied, mth, formula=~classification)))
#  }
}

ord <- ordinate(physeq_rarefied, method="DPCoA", distance=dst)
for (rnk in rank_names(physeq_rarefied)){

    print(plot_ordination(physeq_rarefied, ord, type="taxa", color=rnk )+
      theme(aspect.ratio=1) +
      geom_point(size = 5) +
  #need >3 points to calculate an ellipse
    stat_ellipse() +coord_fixed()+ #geom = "circle"
    geom_path() +coord_fixed()+
      ggtitle(paste("Taxa level", rnk,"-DPCoA Ordination Method"))+ 
        scale_shape_manual(values = c(20,18))+
        theme(plot.title = element_text(size = 16, face = "bold")))
    dev.copy(jpeg,filename=paste0(result_dir,"/plot_Beta_",rnk,"_Taxa", ".jpg"), 
               width=1600, height=800)
    dev.off()
}

```


## plot trees
```{r fig.height=14}
plot_tree(physeq_rarefied_scaled, nodelabf=nodeplotboot(80,0,3), color="short_title",size="abundance", min.abundance=5000, label.tips="short_title",
          base.spacing=0.03,  ladderize="left")#+coord_flip()
dev.copy(jpeg,filename=paste0(result_dir,"/plot_tree_abundance", ".jpg"), 
           width=1500, height=2000)
dev.off()

plot_tree(physeq_rarefied_scaled, color="TimeW", ladderize="left", min.abundance=500) + coord_polar(theta="y")
dev.copy(jpeg,filename=paste0(result_dir,"/plot_tree_TimeW", ".jpg"), 
           width=600, height=600)
dev.off()

```
# DESeq2- taxa differential abundance
importanat!!!!!! this section can be executed in 2 modes:
only for samples that has RNAseq and all samples
###results+PCA

```{r fig.height=10, fig.width=20}
RNA_corr = FALSE
if (RNA_corr){
  md<- meta_16S %>%
  # only patient that has RNAseq
      dplyr::filter(!is.na(RNA_col) & RNA_col!="") %>%
      dplyr::select(PatientID,TimeW,sample_name,batch) %>%
      mutate_if(is.numeric,as.factor) %>%
      arrange(sample_name) %>%
      column_to_rownames("sample_name")
}else{
  md<- meta_16S %>%
  # only patient that has RNAseq
      dplyr::select(patient,classification,sample_name,batch) %>%
      mutate_if(is.numeric,as.factor) %>%
      arrange(sample_name) # %>%
     # column_to_rownames("sample_name")

}
#sum the abundnce per taxa
taxa_count<-physeq_rarefied_data %>%
            dplyr::select(Sample,Abundance,short_title) %>%
    # only samples that has RNAseq
            dplyr::filter(Sample %in% rownames(md)) %>%
            dplyr::group_by(Sample,short_title) %>% 
            dplyr::summarise(total_abund = sum(Abundance))  %>%
            pivot_wider(names_from =Sample, values_from =total_abund)%>%
            replace(is.na(.), 0)%>% 
            dplyr::filter(short_title!="k__Bacteria" & short_title!="d__Bacteria") %>%
            column_to_rownames("short_title")
# filter zero rows
taxa_count <- taxa_count[, colSums(taxa_count != 0) > 0]

if (batch == "303431"){
  dds_taxa <- DESeqDataSetFromMatrix(countData = taxa_count,
                                colData = md,
                                design= ~batch+ PatientID+ TimeW)
}else{
  dds_taxa <- DESeqDataSetFromMatrix(countData = taxa_count,
                                colData = md,
                                design= ~patient+ classification)
}

#dds_taxa$TimeW <- relevel(dds_taxa$TimeW, ref = "0") 
dds_taxa = DESeq(dds_taxa)
resultsNames(dds_taxa)
#[1] "Intercept"                   "classification_post_vs_ctrl" "classification_pre_vs_ctrl" 
#[4] "patient_7_vs_5"              "patient_18_vs_5"       

taxa_abund<-as.data.frame(counts(dds_taxa, normalized=TRUE))

#filter the OTUs using a False Discovery Rate (FDR) cutoff of alpha
alpha = 0
#the significantly differentially abundant OTU between before & after
res_taxa <- results(dds_taxa) # , alpha=alpha
if (!RNA_corr)
  taxa_post_vs_ctrl <- results(dds_taxa, name="classification_post_vs_ctrl") # , alpha=alpha
taxa_pre_vs_ctrl <- results(dds_taxa, name="classification_pre_vs_ctrl") # , alpha=alpha
vst_taxa0 <- varianceStabilizingTransformation (dds_taxa, blind=FALSE)
#vst_taxa0_df<- rownames_to_column(as.data.frame(assay(vst_taxa0)),"short_title")

#if (!grepl("efratsh",getwd())){summary(taxa_3_vs_0)} else {DESeq2::summary(taxa_3_vs_0)}
#if (!grepl("efratsh",getwd())){summary(taxa_6_vs_0)} else {DESeq2::summary(taxa_6_vs_0)}

DESeq2::summary(taxa_pre_vs_ctrl)
DESeq2::summary(taxa_post_vs_ctrl)

if (!RNA_corr){
  #the significantly differentially abundant OTU compare to VHOM
  dds_taxa3<-dds_taxa
  dds_taxa3$TimeW <- relevel(dds_taxa3$TimeW, ref = "3") 
  dds_taxa3 = DESeq(dds_taxa3)
  resultsNames(dds_taxa3)
  taxa_6_vs_3 <- results(dds_taxa3, name="TimeW_6_vs_3") # , alpha=alpha
  vst_taxa3 <- varianceStabilizingTransformation (dds_taxa3, blind=FALSE)
  if (!grepl("efratsh",getwd())){summary(taxa_6_vs_3)} else {DESeq2::summary(taxa_6_vs_3)}
  
  #pairs 
  taxa_all_res <- c("classification_post_vs_ctrl","classification_post_vs_ctrl","Intercept")
}else{  taxa_all_res <- c("taxa_6_vs_0")}
  taxa_all_df <- md %>%
                  dplyr::select(TimeW)

# plot a PCA from vst transformed data
#if (!grepl("efratsh",getwd())){
#  pc<- plot_PCA(dds_TimeW,group=c("TimeW"), set_name="vs_CS", dds_res = summary(res_TimeW))
#} else {
  pc<- plot_PCA(dds_taxa,group=c("classification"), set_name="classification_post_vs_ctrl", dds_res = DESeq2::summary(res_taxa))
#}


ph<- plot_dds_heatmap(dds_taxa,group=c("classification"), set_name="all")

```



###significant pairs analysis , heatmaps and files 

```{r}
#only taxa_6_vs_0
 sig_taxa_all<-taxa_sig_pairs(taxa_thresh=0.58, 
                              taxa_all_results=taxa_all_res,set_name=set)

```
## check contaminantion

```{r}
genus_listk<- subset(taxa_16S_df,short_title %in% sig_taxa_all$short_title) %>%
                  rownames_to_column("OTU")  %>%
                  dplyr::select(Genus) %>%
                  distinct()
gcc<-genus_contaminantion_check(genus_listk)

```


###```{r}
taxa_thresh<-0.58

res_sig_taxa_all<-rownames_to_column(as.data.frame(res_taxa[0,]),"short_title")
first_res<-TRUE
for (res in taxa_all_res){
  #res <- "taxa_6_vs_3"
  print(res)
  res_df<- as.data.frame(get(res))
  
  ## Volcano plot t
  pv <- plot_volcano(res_df,paste0("IBS_16s_taxa_pairs_lfc.",taxa_thresh,"_", res),
                     res_dir = result_dir,
                     lfc=taxa_thresh)

  ## calculate sig taxa
  res_padj_sig <- subset(res_df, padj <= 0.05)
  print(paste(res, "padj sig:",nrow(res_padj_sig)))
  
  
  res_sig <- subset(res_df, pvalue <= 0.05)
  print(paste(res, "pvalue sig:",nrow(res_sig)))
  
  res_sig_taxa <- subset(res_sig, abs(log2FoldChange) >=taxa_thresh)
  print(paste("taxa pvalue_sig:",nrow(res_sig_taxa)))
  taxa_abund_sig<- taxa_abund  %>%
                    rownames_to_column("short_title") %>%
                    dplyr::filter(short_title %in% rownames(res_sig_taxa)) %>%
                    column_to_rownames("short_title") 

  save(res_sig_taxa, taxa_abund_sig, 
       file=paste0(metadata_dir,"/sig.",taxa_thresh,"_taxa_abund_4corr_",res,"_",set, ".Rdata"))

  res_sig_taxa_all<-res_sig_taxa  %>%
                        rownames_to_column("short_title") %>%
                        mutate(pair=res) %>%
                        bind_rows(res_sig_taxa_all)

  
  # LFC barplot
  print(ggplot(rownames_to_column(res_sig_taxa), aes(fill=pvalue, 
                           y=log2FoldChange, 
                           x=reorder(rowname,log2FoldChange)))+
        geom_bar(position="dodge", stat="identity")+
        theme_grey(base_size = 20) +
        ggtitle(paste("barplot taxa sig lfc.",taxa_thresh,res,set))+
        theme(axis.title.y= element_blank())+#, 
        coord_flip())
  dev.copy(jpeg,filename=paste0(result_dir,"/barplot_taxa_sig_lfc.",taxa_thresh,"_",res,"_",set, ".jpg"), 
                 width=1000, height=800)
  dev.off()
  
  #plot sample-OTU heatmap
  st<-str_split_fixed(res,"\\_",4)[,2]
  nd<-str_split_fixed(res,"\\_",4)[,4]
  vst_pair_df<- as.data.frame(assay(get(paste0("vst_taxa",nd))))
  # only the relevant cols
  vst_pair_df<-vst_pair_df[, str_detect(str_split_fixed(colnames(vst_pair_df),"\\_",4)[,3],st)|
                               str_detect(str_split_fixed(colnames(vst_pair_df),"\\_",4)[,3],nd)]
  # only the relevant rows
  taxa_df<-taxa_all_df[str_detect(str_split_fixed(rownames(taxa_all_df),"\\_",4)[,3],st)|
                                     str_detect(str_split_fixed(rownames(taxa_all_df),"\\_",4)[,3],nd),,
                       drop = FALSE]

  
  vst_sig<-vst_pair_df[rownames(res_sig_taxa),]
    if (nrow(vst_sig)>1){
  
    result = tryCatch({
      plot_genes_heatmap(vst_sig,
                         taxa_df,
                         paste0(set,"_",res,"_sig.",taxa_thresh), scl="row",
                         res_dir=result_dir, 
                         plot_title = paste0(set,"_",res,"_sig.",taxa_thresh,"\nscale=row") )
    },
      error = function (condition) {
        plot_genes_heatmap(vst_sig,
                           taxa_df,
                           paste0(set,"_",res,"_sig"), 
                           scl="none",
                           res_dir=result_dir, 
                           plot_title = paste0(set,"_",res,"_sig.",taxa_thresh,"\nscale=none") )
  
    })
  }
} 

ggplot(res_sig_taxa_all, aes(fill=pvalue, 
                           y=log2FoldChange, 
                           x=reorder(short_title,log2FoldChange)))+
        geom_bar(position="dodge", stat="identity")+
        theme_grey(base_size = 20) +
        ggtitle(paste("barplot taxa sig lfc.",taxa_thresh,set))+
        theme(axis.title.y= element_blank())+#, 
        facet_wrap(~pair, strip.position = "top",drop=FALSE, scales = "fixed",nrow = 1)+
        coord_flip()
dev.copy(jpeg,filename=paste0(result_dir,"/barplot_taxa_sig_lfc.",taxa_thresh,"_all_pairs_",set, ".jpg"), 
                 width=1600, height=800)
dev.off()
###```




#correlations taxa-genes - only for 260926

## corelation of taxa/otu between themselves or to RNA sig genes functions

```{r}
cor_sig_taxa_or_OTU <- function(sig_df, set_name,rect=3){
  #sig_df<-taxa_abund_cor
  #set_name<-"OTU"
  cor_sig_OTU <- corr.test(sig_df, method="pearson")
  plot.new()
  #png(file=paste0(result_dir,"/correlation_sig_",set_name,"_",set, ".jpg"), res=300, width=4500, height=4500)
  corrplot::corrplot(cor_sig_OTU$r, p.mat = cor_sig_OTU$p, method = "square",#type = "lower",
                     insig = "label_sig", sig.level = c(.001, .01, .05),
                     cl.pos = 'b', cl.cex = 1, tl.cex = 1,
                     title=paste("\nsignificant",set_name,"correlation", set),
                     pch.cex = .8, pch.col = "white", order="hclust", addrect = rect)
  #wh<-ncol(OTU_abund_sig)*60+100
  #dev.copy(jpeg,filename=paste0(result_dir,"/correlation_sig_OTU_",set, ".jpg"), 
  #         width=wh, height=wh)
  #dev.off()
}

cor_gene_taxa_or_OTU <- function(OTU_abund, set_name){
  #OTU_abund<-w #taxa_abund_all
  #set_name<-"taxa"

  pdf(file=paste0(result_dir,"/Correlations_",set_name,"_gene_lfc.",thresh, ".pdf"),
      paper="USr",width=11.69, height=8.27) 
  wb <- openxlsx::createWorkbook()
  for (fn in file_names){
    #fn<-file_names[1]
    print(fn)
    #in IBS RNA was done only for w6 and w0
    st<-"6"
    nd<-"0"
    
   # if (nd!="CTRL"){ #there is no CTRL for OTU
      #open excel sheet
      sheet_name <- paste(st,"_",nd)
      addWorksheet(wb, sheet_name)

      #print headline
      rects <- data.frame(x = 1:1,
                    colors =  "magenta",
                    text = paste(st,"_vs_",nd))

      print(ggplot(rects, aes(x, y = 0, fill = colors, label = text)) +
              geom_tile(width = .9, height = .9) + # make square tiles
              geom_text(color = "white") + # add white text in the middle
              scale_fill_identity(guide = "none") + # color the tiles with the colors in the data frame
              coord_fixed() + # make sure tiles are square
              theme_void()) # remove any axis markings



      #load gene_count_sig from RNAseq
      load( paste0(metadata_dir,"/../IBS_RNAseq/",file=fn))
      print(paste(st,"-",nd,"no of sig genes:",nrow(dds_count_sig)))   

      load( paste0(metadata_dir,"/sig_abund_4corr_res_6_vs_0_",set,".Rdata"))
      print(paste(st,"-",nd,"no of sig OTUs:",nrow(res_16s_sig_1)))   
      #filter sig accpording to the pair, but different for taxa and OTU
      if (set_name=="OTU"){
          taxa_abund_sig<- OTU_abund %>%
                              rownames_to_column("OTU_taxa") %>%
                              dplyr::filter(OTU %in% rownames(res_16s_sig_1)) %>%
                              dplyr::select(-OTU) %>%
                              column_to_rownames("OTU_taxa") %>%
                              t %>%
                              as.data.frame() %>%
                              rownames_to_column("SampleID") %>%
                              #filter only the samples that were used for RNAseq
                              dplyr::filter(SampleID %in%unique(meta_RNA$ID_16S))#%>%
      }else{
          taxa_abund_sig<- OTU_abund %>%
                              #rownames_to_column("short_title") %>%
                              dplyr::filter(short_title %in% 
                                              str_split_fixed(res_16s_sig_1_taxa$OTU_taxa,"\\-",2)[,1] ) %>%
                              #column_to_rownames("short_title") %>%
                              t %>%
                              as.data.frame() %>%
                              rownames_to_column("SampleID") %>%
                              #filter only the samples that were used for RNAseq
                              dplyr::filter(SampleID %in%unique(meta_RNA$ID_16S))#%>%
        
      }

      # correlating the samples (any_of() to allow missing variables)+filter unwanted genes
      gene_count_sig <- dds_count_sig %>%
                        #rownames_to_column("gene") %>%
                        dplyr::filter(!str_detect(gene,"^Gm")) %>%
                        dplyr::filter(!str_detect(gene,"Rik$")) %>%
                        column_to_rownames("gene") %>%
                        t %>%
                        as.data.frame(stringsAsFactors=FALSE) %>%
                        #convert to 16S sample ID
                        rownames_to_column("SampleID") #%>%
                        #left_join(meta_RNA, by=c("SampleID","sample_name"))
      gene_count_sig<- merge(gene_count_sig,meta_RNA,by.x="SampleID", by.y="sample_name")
      
      #remove columns sum zero
      gene_count_sig<-gene_count_sig[, colSums(gene_count_sig != 0) > 0]
      print(paste(st,"vs",nd,"no of non-zero genes:",ncol(gene_count_sig)))   
      
        # only the relevant cols
      taxa_abund_sig_cor<-taxa_abund_sig %>%
#                        rownames_to_column("RNA_col") %>%
#                        dplyr::filter(RNA_col %in% rownames(gene_count_sig)) %>%
#                        dplyr::select(-SampleID) %>%
                        column_to_rownames("SampleID")
                        #t %>%
                        #as.data.frame()
      #remove columns sum zero
      taxa_abund_sig_cor<-taxa_abund_sig_cor[, colSums(taxa_abund_sig_cor != 0) > 0]
      print(paste(st,"vs",nd,"no of non-zero OTUs:",ncol(taxa_abund_sig_cor))) 
      
      if (ncol(taxa_abund_sig_cor)>8) {
        csto<-cor_sig_taxa_or_OTU(taxa_abund_sig_cor, "OTU", rect = 4)
      } else{
        csto<-cor_sig_taxa_or_OTU(taxa_abund_sig_cor, "OTU")
      } 

      corr_p<-data.frame(gene=character(),taxa=character(),pval=numeric(),stringsAsFactors = FALSE)
      corr_r<-data.frame(gene=character(),taxa=character(),r=numeric(), stringsAsFactors = FALSE)
      for (gene in colnames(subset(gene_count_sig,select=-c(SampleID,ID_16S)))){
        for (taxa in colnames(taxa_abund_sig_cor)){
            #gene<-"Abca13"
            #taxa<-" f__Lachnospiraceae-8843b848d72b39e9cf7173a809bb81c1"
          print(paste("gene:",gene,"taxa:",taxa))
            taxa_abund_sig_cor_p<-subset(taxa_abund_sig_cor,select=taxa)
            #limit the number of zero in OTUs 
            non_zero_otu<-nrow(taxa_abund_sig_cor_p %>%
                    filter_at(vars(contains("__")), any_vars(. > 0)))
            print(paste("non-zero_otu:",non_zero_otu))
            if (non_zero_otu>2){
                gene_count_sig_p<-subset(gene_count_sig,select=c(gene,"ID_16S"))
                gene_taxa<-merge(taxa_abund_sig_cor_p,gene_count_sig_p, by.x="row.names",by.y="ID_16S")
                #corr.test uses cor to find the correlations for either complete or pairwise data and 
                #reports the sample sizes and probability values as well.
                cor_mat_p <- corr.test(gene_taxa[,gene],gene_taxa[,taxa] , method="pearson")
                corr_p<-rbind.data.frame(corr_p,cbind.data.frame(gene=gene,
                                           taxa=taxa,
                                           pval=cor_mat_p$p, 
                                           stringsAsFactors = FALSE),stringsAsFactors = FALSE)
                corr_r<-rbind.data.frame(corr_r,cbind.data.frame(gene=gene,
                                           taxa=taxa,
                                           r=cor_mat_p$r, 
                                           stringsAsFactors = FALSE),stringsAsFactors = FALSE)
      
                #plot correlation regresion line for the significants
                if (cor_mat_p$p<0.01) {
                  print(paste("cor_mat_p$p<0.01 gene:",gene,"taxa:",taxa))
      
                  gg<-ggscatter(gene_taxa, x = gene, y = taxa, 
                              main=paste("\n\n\nTaxa- gene correlation in ", st,"vs", nd,
                                         "\npval:",cor_mat_p$p,"r:",cor_mat_p$r),
                              font.subtitle = c(10,  "blue"),
                              add = "reg.line", conf.int = TRUE, 
                              cor.coef = TRUE, cor.method = "pearson",
                              xlab = gene, ylab = taxa,
                              label = "Row.names", repel = TRUE)
                  text<-tableGrob(gene_taxa)
                  together <- arrangeGrob(text, gg, ncol = 1,nrow = 2)
                  plot(together)
                } 
            }
        }
      } 
      #print(corr_p)
      #print(corr_r)
      #write sig cor genes per OTU to excel
      corr_p_sig<-subset(corr_p,pval<0.1,select=-pval)
      print(paste("no of cor pval<0.1:",nrow(corr_p_sig)))
      colno<-1
      for (t in unique(corr_p_sig$taxa)){
        #t<-" s__subflava" 
        corr_p_sig_t<-subset(corr_p_sig,taxa==t,select=gene)
        colnames(corr_p_sig_t)<-t
        writeData(wb, sheet = sheet_name,corr_p_sig_t , rowNames = F, startCol = colno)
        colno=colno+1
      }


      for(t in unique(corr_r$taxa)){
        corr_r_4plot<-corr_r %>%                   
                      mutate(gene=as.factor(gene)) %>%
                      dplyr::filter(taxa==t) %>%
                      arrange( r)
        print(ggplot(corr_r_4plot, aes(x=reorder(gene,r), y=r)) + 
                geom_point() +
                scale_y_continuous()+
                ggtitle(paste(st,"_",nd," correlation for OTUs", t)))
      }
      
      #spread (move rows into columns).
      corr_p_spread<- spread(corr_p, key=gene, value=pval, convert = FALSE) %>% 
                        column_to_rownames("taxa") %>%
                        as.matrix()
      corr_r_spread<- spread(corr_r, key=gene, value=r, convert = FALSE) %>% 
                        column_to_rownames("taxa")  %>%
                        as.matrix()
  
      #print(corr_p_spread)
      #print(corr_r_spread)
      corrplot::corrplot(corr_r_spread, p.mat = corr_p_spread, method = "square",
               insig = "label_sig", sig.level = c(.001, .01, .05),cl.pos = 'b',
               title=paste("\n\n\nOTU-significant genes correlation", st,"vs", nd,
                           "\nno of genes:",ncol(gene_count_sig)-2,
                           "\nno of OTUs:",ncol(taxa_abund_cor)),
               tl.cex=.5,pch.cex = .6, pch.col = "white") #, order="FPC"
  
        
    }
  #}
  dev.off()
  openxlsx::saveWorkbook(wb, file=paste0(result_dir,"/correlation_sig_genes_lfc.",
                                         thresh,"_OTUs_",set_name,"_", set, ".xlsx"), 
                       overwrite = TRUE)

}
```
## corelation of OTU/taxa(mean OTU) to RNA sig genes

```{r}
#read RNAseq metadata for converting to the same sample ID
meta_RNA <- as.data.frame(read_excel(paste0(metadata_dir,"/../IBS_RNAseq/IBS_metadata.xlsx")),
                          stringsAsFactors=FALSE)
#remove sample 384
meta_RNA <- meta_RNA %>%
              dplyr::filter(sample!="N384") %>%   #& sample!="N389"
              dplyr::select(c("sample_name","ID_16S"))
              #column_to_rownames("sample")

#load RNAseq sig genes
load( paste0(metadata_dir,"/sig_abund_4corr_",set, ".Rdata"))
thresh<-1
file_names <-list.files(paste0(metadata_dir,"/../IBS_RNAseq/"), 
                        pattern= glob2rx(paste0("*_sig.",thresh,".Rdata$")),
                        recursive = TRUE)#,full.names =TRUE)

#taxa level (OTU mean per taxa)
taxa_abund_all_mean<- dds_abund %>%
                    as.data.frame(stringsAsFactors=FALSE)  %>%
                    rownames_to_column("OTU") %>%
                    #dplyr::filter(OTU %in% rownames(res_sig_pval_lfc.1_TimeW)) %>%
                    left_join(dplyr::select(rownames_to_column(taxa_16S_df,"OTU"),c(OTU,short_title)),
                              by="OTU") %>%
                    #gather(key=sample_name,value=abund, -c(short_title, OTU)) %>%
                    dplyr::select(-OTU) %>%
                    #pivot_wider(names_from =short_title, values_from =abund)
                    group_by(short_title) %>%
                    summarise_each(mean) %>%
                    column_to_rownames("short_title") #%>%
#taxa_abund_all_p<-subset(taxa_abund_all, str_detect(short_title," f__"))
#w<-taxa_abund_all_p%>%
#  group_by(short_title) %>%
#  mutate(row = row_number()) %>%
#  pivot_wider(id_cols = c(row, sample_name), names_from =short_title, values_from =abund)%>%
#  dplyr::select(-row) 


#dcast.data.table(as.data.table(taxa_abund_all_p), OTU+sample_name ~ short_title,value.var="abund",drop=T)

cgto<-cor_gene_taxa_or_OTU (taxa_abund_all_mean, "taxa")



#OTU level

## add taxa short name to OTU
OTU_abund_sig<- dds_abund %>%
                    as.data.frame(stringsAsFactors=FALSE)  %>%
                    rownames_to_column("OTU") %>%
                    left_join(dplyr::select(rownames_to_column(taxa_16S_df,"OTU"),c(OTU,short_title)),
                              by="OTU") %>%
                    mutate(OTU_taxa=paste0(short_title,"-",OTU)) %>%
                    dplyr::select(-c(short_title)) %>%
                    column_to_rownames("OTU_taxa") #%>%
#                    t %>%
#                    as.data.frame()

cgto<-cor_gene_taxa_or_OTU (OTU_abund_sig, "OTU")

```


## corelation of taxa(all OTUs) to RNA sig genes

```{r}
corr_sig_taxa_only<- FALSE
taxa_thresh<-0.58
gene_thresh<-1
if (!exists("mouse_ensembl")) mouse_ensembl = useDataset("mmusculus_gene_ensembl",mart=useMart("ensembl"))

#read RNAseq metadata for converting to the same sample ID
meta_RNA <- as.data.frame(read_excel(paste0(metadata_dir,"/../IBS_RNAseq/IBS_metadata.xlsx")),
                          stringsAsFactors=FALSE)
#remove sample 384
meta_RNA <- meta_RNA %>%
              dplyr::filter(sample!="N384") %>%   #& sample!="N389"
              dplyr::select(c("sample_name","ID_16S"))
              #column_to_rownames("sample")

file_names <-list.files(paste0(metadata_dir,"/../IBS_RNAseq/"), 
                        pattern= glob2rx(paste0("*_sig.",gene_thresh,".Rdata$")),
                        recursive = FALSE)#,full.names =TRUE)


#test
pdf(file=paste0(result_dir,"/Correlations_taxa",classif,"_all_gene_lfc.",taxa_thresh, ".pdf"),
    paper="USr",width=11.69, height=8.27) 
wb <- openxlsx::createWorkbook()
#for (fn in file_names){
fn<-file_names[1]
print(fn)
#in IBS RNA was done only for w6 and w0
st<-"6"
nd<-"0"

# if (nd!="CTRL"){ #there is no CTRL for OTU
  #open excel sheet
sheet_name <- paste(st,"_",nd)
addWorksheet(wb, sheet_name)

#print headline
rects <- data.frame(x = 1:1,
              colors =  "magenta",
              text = paste(st,"_vs_",nd))

print(ggplot(rects, aes(x, y = 0, fill = colors, label = text)) +
        geom_tile(width = .9, height = .9) + # make square tiles
        geom_text(color = "white") + # add white text in the middle
        scale_fill_identity(guide = "none") + # color the tiles with the colors in the data frame
        coord_fixed() + # make sure tiles are square
        theme_void()) # remove any axis markings



#load dds_count_sig from RNAseq RNAseq sig genes

load( paste0(metadata_dir,"/../IBS_RNAseq/",file=fn))
print(paste(st,"-",nd,"no of sig genes:",nrow(dds_count_sig))) 


load( paste0(metadata_dir,"/sig.",taxa_thresh,"_taxa_abund_4corr_taxa_6_vs_0_",set, ".Rdata"))

      
if (corr_sig_taxa_only){
  print(paste(st,"-",nd,"no of sig taxa 4corr:",nrow(res_sig_taxa)))   
  res_taxa_4corr<- res_sig_taxa %>%
                arrange(log2FoldChange)
  taxa_abund_4corr<- taxa_abund_sig  %>%
                    t()  %>%
                    as.data.frame(stringsAsFactors=FALSE)  %>%
                    rownames_to_column("sample_name") 
} else{
  print(paste(st,"-",nd,"no of taxa 4corr:",nrow(res_taxa)))   
  res_taxa_4corr<- res_taxa_pval %>%
                arrange(log2FoldChange)
  taxa_abund_4corr<- taxa_abund_all  %>%
                    t()  %>%
                    as.data.frame(stringsAsFactors=FALSE)  %>%
                    rownames_to_column("sample_name") 
  
}


#dcast.data.table(as.data.table(taxa_abund_all_p), sample_name+row ~ short_title,value.var="abund",drop=T)
  gene_count_sig <- dds_count_sig %>%
                    #rownames_to_column("gene") %>%
                    dplyr::filter(!str_detect(gene,"^Gm")) %>%
                    dplyr::filter(!str_detect(gene,"Rik$")) %>%
                    column_to_rownames("gene") %>%
                    t %>%
                    as.data.frame(stringsAsFactors=FALSE) %>%
                    #convert to 16S sample ID
                    rownames_to_column("SampleID") #%>%
                    #left_join(meta_RNA, by=c("SampleID","sample_name"))
  gene_count_sig<- merge(gene_count_sig,meta_RNA,by.x="SampleID", by.y="sample_name")
  
  #remove columns sum zero
  gene_count_sig<-gene_count_sig[, colSums(gene_count_sig != 0) > 0]
  print(paste(st,"vs",nd,"no of non-zero genes:",ncol(gene_count_sig)))   
  
    # only the relevant cols
  taxa_abund_cor<-taxa_abund_4corr %>%
                        mutate_if(is.numeric, ~replace(., is.na(.), 0))
#                        rownames_to_column("RNA_col") %>%
#                        dplyr::filter(RNA_col %in% rownames(gene_count_sig)) %>%
#                        dplyr::select(-SampleID) %>%
                   # column_to_rownames("sample_name")
                    #t %>%
                    #as.data.frame()
  #remove columns sum zero
  taxa_abund_cor<-taxa_abund_cor[, colSums(taxa_abund_cor != 0) > 0]
  print(paste(st,"vs",nd,"no of non-zero OTUs:",ncol(taxa_abund_cor))) 
  

  corr_p<-data.frame(gene=character(),taxa=character(),pval=numeric(),stringsAsFactors = FALSE)
  corr_r<-data.frame(gene=character(),taxa=character(),r=numeric(), stringsAsFactors = FALSE)
  for (gene in colnames(subset(gene_count_sig,select=-c(SampleID,ID_16S)))){
    for (taxa in colnames(subset(taxa_abund_cor,select=-sample_name))){
        #gene<-"Abca13"
        #taxa<-" f__Lachnospiraceae-8843b848d72b39e9cf7173a809bb81c1"
      print(paste("gene:",gene,"taxa:",taxa))
        taxa_abund_cor_p<-subset(taxa_abund_cor,select=c(taxa,"sample_name"))
        #limit the number of zero in OTUs 
        non_zero_otu<-nrow(taxa_abund_cor_p %>%
                filter_at(vars(contains("__")), any_vars(. > 0)))
        print(paste("non-zero_otu:",non_zero_otu))
        if (non_zero_otu>2){
            gene_count_sig_p<-subset(gene_count_sig,select=c(gene,"ID_16S"))
            gene_taxa<-merge(taxa_abund_cor_p,gene_count_sig_p, by.x="sample_name",by.y="ID_16S")
            #corr.test uses cor to find the correlations for either complete or pairwise data and 
            #reports the sample sizes and probability values as well.
            cor_mat_p <- corr.test(gene_taxa[,gene],gene_taxa[,taxa] , method="pearson")
            corr_p<-rbind.data.frame(corr_p,cbind.data.frame(gene=gene,
                                       taxa=taxa,
                                       pval=cor_mat_p$p, 
                                       stringsAsFactors = FALSE),stringsAsFactors = FALSE)
            corr_r<-rbind.data.frame(corr_r,cbind.data.frame(gene=gene,
                                       taxa=taxa,
                                       r=cor_mat_p$r, 
                                       stringsAsFactors = FALSE),stringsAsFactors = FALSE)
  
            #plot correlation regresion line for the significants
            #if (cor_mat_p$p<0.01) {
            #  print(paste("cor_mat_p$p<0.01 gene:",gene,"taxa:",taxa))
  
             # gg<-ggscatter(gene_taxa, x = gene, y = taxa, 
            #              main=paste("\n\n\nTaxa- gene correlation in ", st,"vs", nd,
            #                         "\npval:",cor_mat_p$p,"r:",cor_mat_p$r),
            #              font.subtitle = c(10,  "blue"),
            #              add = "reg.line", conf.int = TRUE, 
            #              cor.coef = TRUE, cor.method = "pearson",
            #              xlab = gene, ylab = taxa,
            #              label = "Row.names", repel = TRUE)
            #  text<-tableGrob(gene_taxa)
            #  together <- arrangeGrob(text, gg, ncol = 1,nrow = 2)
            #  plot(together)
            #} 
        }
    }
  } 
  #print(corr_p)
  #print(corr_r)
cgtx<- corr_gene_tax(corr_p,corr_r, gene_res_pval_sig, res_taxa_4corr, wb, sheet_name,
                     in_tl_size=.5)
dev.off()
openxlsx::saveWorkbook(wb, file=paste0(result_dir,"/correlation_sig.",gene_thresh,
                                           "_genes_taxa_all_", set,classif, ".xlsx"), 
                           overwrite = TRUE)

# compare number of genes corellated in each 2 bacterias
for (t1 in unique(corr_p$taxa)) {
  for (t2 in unique(corr_p$taxa)) {
    if (t1!=t2){
      print(paste("t1:",t1,"t2:",t2))
      t1_df<- subset(corr_p, pval<=0.05 & taxa==t1)
      t2_df<- subset(corr_p, pval<=0.05 & taxa==t2)
      venn_list<-list( t1_df$gene, t2_df$gene)
      names(venn_list)<-c(t1,t2)
      venn(venn_list)
    }
  }
}

```

#correlation taxa-integrity

```{r}
meta_int<-meta_16S %>%
            dplyr::filter(!is.na(integrity24))
taxa_int<- physeq_rarefied_scaled_data %>%
            dplyr::select(Sample,Abundance,short_title) %>%
            dplyr::group_by(Sample,short_title) %>% 
            dplyr::summarise(total_abund = sum(Abundance))  %>%
    # only samples that has integrity
            right_join(dplyr::select(meta_int,sample_name,integrity24),by=c("Sample"="sample_name"))
pdf(file=paste0(result_dir,"/Correlations_",set,"_taxa_integrity", ".pdf"),
      paper="USr",width=11.69, height=8.27) 
corr_sig<-data.frame(taxa=character(),pval=numeric(), r=numeric(),stringsAsFactors = FALSE)
for (st in unique(taxa_int$short_title)){
  #st<- unique(taxa_int$short_title)[1]
  print(paste("taxa:",st))
  taxa_int_p<-subset(taxa_int,short_title==st)
  cor_mat_p <- corr.test(taxa_int_p$total_abund,taxa_int_p$integrity24 , method="pearson")
  if (!is.na(cor_mat_p$p)) { #        
  #if (cor_mat_p$p<0.1 & !is.na(cor_mat_p$p)) {
      print(paste("cor_mat_p$p:",cor_mat_p$p,"taxa:",st))

      gg<-ggscatter(taxa_int_p, x ="total_abund"  , y = "integrity24", 
                    main=paste("\n\n\nTaxa- integrity correlation in taxa:", st,
                               "\npval:",cor_mat_p$p,"r:",cor_mat_p$r),
                    font.subtitle = c(10,  "blue"),
                    add = "reg.line", conf.int = TRUE, 
                    cor.coef = TRUE, cor.method = "pearson",
                    label = "Sample", repel = TRUE)
        text<-tableGrob(taxa_int_p)
        together <- arrangeGrob(text, gg, ncol = 1,nrow = 2)
        plot(together)
        if (cor_mat_p$p<0.1) {
          corr_sig<-rbind.data.frame(corr_sig,cbind.data.frame(taxa=st,
                                     pval=cor_mat_p$p, 
                                     r=cor_mat_p$r, 
                                    stringsAsFactors = FALSE),stringsAsFactors = FALSE)
         }
    }
  } 
  #print(corr_p)
  #print(corr_r)
dev.off()


```

#Clustering

```{r}
#dds_count_sig.1 <- subset(dds_t_count, rownames(dds_count) %in% row.names(sig.1))

# cluster RLD for all vs CTRL sig.1

cluster_set <- "vsd.1"
#dds_count_sig.1/vsd_sig.1_df
#count_df<-vsd_sig.1_df
count_df<- vsd_sig.1_df %>%
                    rownames_to_column("OTU") %>%
                    left_join(dplyr::select(rownames_to_column(taxa_16S_df,"OTU"),c(OTU,short_title)),
                              by="OTU") %>%
                    mutate(OTU_taxa=paste0(short_title,"-",OTU)) %>%
                    dplyr::select(-c(OTU,short_title)) %>%
                    column_to_rownames("OTU_taxa") #%>%

# Before clustering, scaling the data to identify clusters of genes that share similar expression profiles rather than similar expression levels.
count_df_scaled <- t(scale(t(count_df))) # Centers and scales data.
#search for K
pk<-plot_k(count_df_scaled,"all vs CTRL",max_k=5)

kmeans_cluster <- kmeans(count_df_scaled, 3)
kmeans_cluster$centers
fviz_cluster(kmeans_cluster, data = count_df_scaled, main = "OTU kmeans k=3")
dev.copy(jpeg,filename=paste0(result_dir,"/kmeans_",set, ".jpg"), 
            width=1200, height=600)
dev.off()

```



#Picrust2
```{r}
func_abund<- openxlsx::read.xlsx(paste0(metadata_dir,"/exports/path_abun_unstrat_descrip.xlsx"))
rownames(func_abund)<-NULL
md<- meta_16S %>%
  # only patient that has RNAseq
      dplyr::select(PatientID,TimeW,sample_name,batch) %>%
      mutate_if(is.numeric,as.factor) %>%
      arrange(sample_name) %>%
      column_to_rownames("sample_name")
func_all_df <- md %>%
                  dplyr::select(TimeW)

func_abund_df <- func_abund %>% 
              as.data.frame(stringsAsFactors=FALSE)  %>%
              dplyr::select(-description)   %>%
              column_to_rownames("pathway") %>% 
              t()  %>% 
              as.data.frame(stringsAsFactors=FALSE)  %>%
              rownames_to_column("SampleID")  %>%
  #special case- remove unwanted sample
              mutate(SampleID=str_replace(SampleID,"260926-IBS","IBS"))  %>% # special case
              mutate(SampleID=str_replace(SampleID,"260916","IBS"))  %>% # special case
              mutate(SampleID=str_replace(SampleID,"303431-",""))  %>% # special case
              mutate(SampleID=str_replace_all(SampleID,"\\-","_"))  %>% # dot to underscore
              #mutate(SampleID=meta_16S[match(SampleID,meta_16S$SampleID), "sample_name"]) %>% #+TimeW+PatientID
              left_join(dplyr::select(meta_16S,c(SampleID,sample_name)),by="SampleID") %>% 
              dplyr::select(-SampleID) %>% 
              #dplyr::mutate_if(is.numeric, ~ . * 1000)  %>% # *1000 as deseq2 works with integers
              dplyr::mutate_if(is.numeric, as.integer)   %>% #deseq2 works with integers
              arrange(factor(sample_name,levels=rownames(md))) %>%
              column_to_rownames("sample_name") %>% 
              t()  %>% 
              as.data.frame(stringsAsFactors=FALSE)  


dds_func <- DESeqDataSetFromMatrix(countData = func_abund_df,
                                colData = md,
                                design=~ PatientID+TimeW) #
dds_func$TimeW <- relevel(dds_func$TimeW, ref = "0") 
dds_func<- DESeq(dds_func)
dds_func<- estimateSizeFactors(dds_func)
plotDispEsts(dds_func)
resultsNames(dds_func)


func_3_vs_0 <- results(dds_func, name="TimeW_3_vs_0") # , alpha=alpha
DESeq2::summary(func_3_vs_0)
func_6_vs_0 <- results(dds_func, name="TimeW_6_vs_0") # , alpha=alpha
DESeq2::summary(func_6_vs_0)
vst_func0 <- varianceStabilizingTransformation (dds_func, blind=FALSE)

dds_func3<-dds_func
dds_func3$TimeW <- relevel(dds_func3$TimeW, ref = "3") 
dds_func3 = DESeq(dds_func3)
resultsNames(dds_func3)
func_6_vs_3 <- results(dds_func3, name="TimeW_6_vs_3") # , alpha=alpha
DESeq2::summary(func_6_vs_3)
vst_func3 <- varianceStabilizingTransformation (dds_func3, blind=FALSE)
  
#pairs 
func_all_res <- c("func_3_vs_0","func_6_vs_0","func_6_vs_3")


pc<- plot_PCA(dds_func,group=c("TimeW"), set_name="vs_0", dds_res = DESeq2::summary(results(dds_func)))

ph<- plot_dds_heatmap(dds_func,group=c("TimeW"), set_name="all")

#
sig_func_all<-taxa_sig_pairs(taxa_thresh=0.58, 
                              taxa_all_results=func_all_res,set_name=set, use_padj=F, corr=F)
sig_func_all_desc<-sig_func_all %>%
                      dplyr::rename(pathway=short_title) %>%
                      left_join(dplyr::select(func_abund,c(pathway,description)),
                                              by="pathway") %>%
                      arrange(pathway)
ggtexttable(unique(dplyr::select(sig_func_all_desc,c(pathway,description))), rows = NULL)
dev.copy(jpeg,filename=paste0(result_dir,"/Function_description_sig_lfc.",taxa_thresh,"_all_pairs_",set, ".jpg"), 
            width=600, height=nrow(sig_func_all_desc)*15)
dev.off()

```


#Burrito
```{r}
func_abund<- openxlsx::read.xlsx(paste0(metadata_dir,"/burrito_function_abundances.xlsx"))
rownames(func_abund)<-NULL
func_abund_df <- func_abund %>% 
              as.data.frame(stringsAsFactors=FALSE)  %>%
              column_to_rownames("SubPathway") %>% 
              t()  %>% 
              as.data.frame(stringsAsFactors=FALSE)  %>%
              rownames_to_column("SampleID")  %>%
  #special case- remove unwanted sample
              mutate(SampleID=str_replace(SampleID,"260926-IBS","IBS"))  %>% # special case
              mutate(SampleID=str_replace(SampleID,"260916","IBS"))  %>% # special case
              mutate(SampleID=str_replace_all(SampleID,"\\-","_"))  %>% # dot to underscore
              #mutate(SampleID=meta_16S[match(SampleID,meta_16S$SampleID), "sample_name"]) %>% #+TimeW+PatientID
              left_join(dplyr::select(meta_16S,c(SampleID,sample_name)),by="SampleID") %>% 
              dplyr::select(-SampleID) %>% 
              dplyr::mutate_if(is.numeric, ~ . * 1000)  %>% # *10000 as deseq2 works with integers
              dplyr::mutate_if(is.numeric, as.integer)   %>%
              arrange(factor(sample_name,levels=rownames(md))) %>%
              column_to_rownames("sample_name") %>% 
              t()  %>% 
              as.data.frame(stringsAsFactors=FALSE)  


dds_func <- DESeqDataSetFromMatrix(countData = func_abund_df,
                                colData = md,
                                design=~ TimeW) #PatientID+
dds_func$TimeW <- relevel(dds_func$TimeW, ref = "0") 
dds_func<- DESeq(dds_func)
dds_func<- estimateSizeFactors(dds_func)
dds_func<- estimateDispersionsGeneEst(dds_func)
dispersions(dds_func) <- mcols(dds_func)$dispGeneEst
plotDispEsts(dds_func)
resultsNames(dds_func)
func_After_vs_Before <- results(dds_func) # , alpha=alpha
vst_funcBefore <- varianceStabilizingTransformation (dds_func, blind=FALSE)

if (!grepl("efratsh",getwd())){summary(func_After_vs_Before)} else {DESeq2::summary(func_After_vs_Before)}

pc<- plot_PCA(dds_func,group=c("TimeW"), set_name="vs_0", dds_res = DESeq2::summary(func_After_vs_Before))

ph<- plot_dds_heatmap(dds_func,group=c("TimeW"), set_name="all")

func_After_vs_Before_sig<-  subset(as.data.frame(func_After_vs_Before), pvalue <= 0.05)

func_all_res <- "func_After_vs_Before"
#only taxa_After_vs_Before
 sig_func_all<-taxa_sig_pairs(taxa_thresh=0.58, 
                              taxa_all_results=func_all_res,set_name=set)

```

# FishTaco - prepare data
```{r}
otu<- read.table(paste0(metadata_dir,"/exports",classif,"/otu_table.txt"), header = T,  sep= "\t")
fishTaco<- otu %>%
              as.data.frame(stringsAsFactors=FALSE)  %>%
              column_to_rownames("OTU.ID") %>% 
              t()  %>% 
              as.data.frame(stringsAsFactors=FALSE)  %>%
              rownames_to_column("SampleID")  %>%
              mutate(SampleID=str_replace(SampleID,"X",""))  %>% # remove prefix
              mutate(SampleID=str_replace_all(SampleID,"\\.","-"))  %>% # replace dot to -
              column_to_rownames("SampleID") %>%
              t()  %>% 
              as.data.frame(stringsAsFactors=FALSE)  %>%
              rownames_to_column("OTU.ID")  %>%
              left_join(dplyr::select(rownames_to_column(taxa_16S_df,"OTU.ID"),
                                      c("OTU.ID","short_title")),by="OTU.ID") %>%
              dplyr::select(-OTU.ID) 
openxlsx::write.xlsx (fishTaco, paste0(metadata_dir,"/taxa_fishTaco.xlsx"),
                              sheetName ="fishTaco" ,col.names = TRUE,row.names=F) 

```


